File: 48.js
// LeetCode 48 Rotate imgae
// Reference: https://leetcode.com/problems/rotate-image/solutions/1175496/js-python-java-c-easy-4-way-swap-solution-w-explanation/?envType=study-plan-v2&envId=top-interview-150

/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    
    let n = matrix.length, depth = ~~(n / 2)
    for (let i = 0; i < depth; i++){

        let len = n - 2 * i - 1, opp = n - 1 - i

        for (let j = 0; j < len; j++){
            let temp = matrix[i][i + j]
            matrix[i][i + j] = matrix[opp - j][i]
            matrix[opp - j][i] = matrix[opp][opp - j]
            matrix[opp][opp - j] = matrix[i + j][opp]
            matrix[i + j][opp] = temp
        }
    }
};

File: 48.java
// LeetCode 48 Rotate imgae
// Reference: https://leetcode.com/problems/rotate-image/solutions/18879/ac-java-in-place-solution-with-explanation-easy-to-understand/?envType=featured-list&envId=top-interview-questions?envType=featured-list&envId=top-interview-questions

class Solution {
    public void rotate(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++){
            for (int j = i; j < matrix[0].length; j++){
                int temp = 0;
                temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;

            }
        }

        for (int i = 0; i < matrix.length; i++){
            for (int j = 0; j < matrix.length / 2; j++){
                int temp = 0;
                temp = matrix[i][j];
                matrix[i][j] = matrix[i][matrix.length - 1 - j];
                matrix[i][matrix.length - 1 - j] = temp;
            }
        }
    }
}